hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/BackpressureRegulator.java

```diff
@@ -178,25 +219,93 @@ boolean isSyncForced(BackupAwareOperation backupAwareOp) {
             return false;
         }
 
-        for (; ; ) {
-            int current = syncCountdown.decrementAndGet();
-            if (current > 0) {
-                return false;
+        if (currentTimeMillis() > timeoutMs) {
+            // no overload is detected.
+            return false;
+        }
+
+        return random.nextInt(HUNDRED) <= asyncToSyncBackupRatio;
+    }
+
+    public void start() {
+        if (asyncBackupOverloadCheckIntervalMs > 0) {
+            asyncBackupOverloadDetectionThread.start();
+        }
+    }
+
+    public void shutdown() {
+        asyncBackupOverloadDetectionThread.shutdown();
+    }
+
+    private final class AsyncBackupOverloadDetectionThread extends Thread {
+        private volatile boolean stop;
+
+        private AsyncBackupOverloadDetectionThread(String hzName) {
+            super(createThreadName(hzName, "AsyncBackupOverloadDetectionThread"));
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (!stop) {
+                    sleep();
+
+                    long startMs = System.currentTimeMillis();
+                    long maxBytesPending = maxBytesPending();
+                    if (maxBytesPending < asyncBackupLowWaterMark) {
+                        continue;
+                    }
+
+                    // overload detected.
+                    timeoutMs = Clock.currentTimeMillis() + asyncBackupBackpressureWindowMs;
+                    maxBytesPending = min(maxBytesPending, asyncBackupHighWaterMark);
+                    asyncToSyncBackupRatio = Math.round(HUNDRED * ((float) maxBytesPending) / asyncBackupHighWaterMark);
+                    long duration = System.currentTimeMillis() - startMs;
+                    logger.info("ratio:" + asyncToSyncBackupRatio + " max:" + maxBytesPending + " duration:" + duration + " ms");
+                }
+            } catch (Throwable t) {
+                logger.severe("Overload DetectionThread failed", t);
             }
+        }
 
-            if (syncCountdown.compareAndSet(current, randomSyncDelay())) {
-                return true;
+        private void sleep() {
+            try {
+                Thread.sleep(asyncBackupOverloadCheckIntervalMs);
+            } catch (InterruptedException e) {
+                // we can ignore it since we check in a loop if this thread needs to be stopped.
```

Comment:
I'm aware that this PR is most likely stale, but I'm trying to figure out what was happening in development during the last months and I have a single question out of pure curiosity.

> // we can ignore it since we check in a loop if this thread needs to be stopped.

It's true but the interruption gets lost while the loop keeps looping. Shouldn't the interruption flag be restored in the `catch{}` and then the `while{}` condition check for either `!stop` or if the thread was interrupted?

but is it ok to